Refacto AdminImportControllerCore

LittleBigDev opened the pull request:
<!-- Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information: -->

| Questions     | Answers                                                                          |
|---------------|----------------------------------------------------------------------------------|
| Branch?       | develop                                                                          |
| Description?  | Refactoring AdminImportControllerCore code. |
| Type?         |improvement|
| Category?     |BO|
| BC breaks?    | no                                                                               |
| Deprecations? | no                                                                               |
| Fixed ticket? | N/A                          |
| How to test?  | N/A                |

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8122)
<!-- Reviewable:end -->

LittleBigDev commented:
To reviewers : bug was fixed by `return $line_count` in `AdminImportController.php` line 3701. Everything else is code cleaning / refactoring.

xBorderie commented:
I guess if you have to explain it, that means it would have been easier to make at least two commits: one for the fix, the other for the cleaning/refactoring :)

LittleBigDev commented:
Well, I thought it wasn't worth it committing a single line change, then the rest... Should I fix this ?

xBorderie commented:
If it were up to me, I'd do:
- 1 commit for the fix
- 1 commit for the refactoring
- 1 commit for the code cleaning

This way, no commit pollutes the other.

But I'm not a Core dev, so... :)

LittleBigDev committed:
BO: Some refactoring into store contacts csv import code

LittleBigDev committed:
CO: PSR compliancy on AdminImportController; lines length and variable naming.

LittleBigDev commented:
Fixed :)

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -3682,30 +3697,49 @@ public function storeContactImport($offset = false, $limit = false, $validateOnl
-            if (!(AdminImportController::copyImg($store->id, null, $store->image, 'stores', !$regenerate))) {
-                $this->warnings[] = $store->image.' '.$this->trans('cannot be copied.', array(), 'Admin.Advparameters.Notification');
+        if (!empty($store->image)) {
+            $imgWasCopied = AdminImportController::copyImg($store->id, null, $store->image, 'stores', !$regenerate);
```
If the value is not used on another line, no need to store it in a variable. The previous version was correct in my opinion.

LittleBigDev replied:
It was done to enhance readability of the code. `AdminImportController::copyImg()` return value was unclear.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -3682,30 +3697,49 @@ public function storeContactImport($offset = false, $limit = false, $validateOnl
+            $imgWasCopied = AdminImportController::copyImg($store->id, null, $store->image, 'stores', !$regenerate);
+            if (!$imgWasCopied) {
+                $this->warnings[] = $store->image . ' '
+                    . $this->trans('cannot be copied.', [], 'Admin.Advparameters.Notification');
```
ping @AlexEven 

If we mix refacto and bug fix in this PR, it would be great to make `$store->image` as a variable in the translated message. It may be not available on the beginning of the string in some language. However, this means we cannot merge it for 1.7.2.x if 1.7.2.0 is released.

xBorderie replied:
I confirm that we'd rather have 
`'%1 cannot be copied` (or some equivalent) 
than 
`$img . ' ' . ' cannot be copied'`.

xBorderie replied:
And yeah, if the PR introduces string changes, it's kinda late in the 1.7.2.x game, and might force the PR to be moved to 1.7.3.x...

LittleBigDev replied:
I'd prefer too but it would be outside BOOM-294's scope.
I'm ok to fix this in another PR (this is not the only one that would need a fix).

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -3682,30 +3697,49 @@ public function storeContactImport($offset = false, $limit = false, $validateOnl
+        $storeId = isset($info['id']) ? (int)$info['id'] : null;
+        $store   = new Store($storeId);
+
+        $store->force_id = (bool)$forceIds;
```
What happen for `$force_ids` and `Store::existsInDatabase(...)`?

You should explain why you changed the tests, because after a quick look this is unclear. If we want to refacto it without modifying the test cases, I would have written i.e:

```php
if (!empty($info['id']) && ($force_ids || Store::existsInDatabase((int)$info['id'], 'store'))) {
       $storeId = (int)$info['id'];
} else {
       $storeId = null;
}
```

LittleBigDev replied:
All this code block is redundant with `Store` and `ObjectModel` constructors' logic :
- passing empty or unknown id will lead to an unloaded `Store` object
- `$force_ids` is used only when `Store::add()` is executed and has no meaning when we need to decide whether we pass an id to the constructor or not.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -3762,44 +3797,66 @@ public function storeContactImportOne($info, $shop_is_feature_active, $regenerat
+
+        // If errors, log them and stop execution
+        if (true !== $fieldsValidationResult || true !== $langFieldsValidationResult) {
+            $errorGenericMessage = $this->trans('Store is invalid', [], 'Admin.Advparameters.Notification');
```
`[]` --> `array()`

maximebiloe replied:
Don't hesitate to look at http://build.prestashop.com/news/prestashop-coding-standards/

LittleBigDev replied:
Prestashop coding standards didn't help me on this one. Symfony either. I used the short notation to enhance readability.
If it really grinds your gears, I can revert this...

Quetzacoalt91 replied:
It does not grind our gears like for Peter Griffin, but the rules are the same for everybody. :wink:

LittleBigDev replied:
We should write down in these rules that short array notation is forbidden. It's not the case right now.

eternoendless replied:
After discussion, we settled for `array()`

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -3762,44 +3797,66 @@ public function storeContactImportOne($info, $shop_is_feature_active, $regenerat
                 );
-            }
-        } else {
-            $this->errors[] = $this->trans('Store is invalid', array(), 'Admin.Advparameters.Notification').' ('.$store->name.')';
```
@AlexEven, aren't we supposed to add `('.$store->name.')` in the translated string as a variable? This could be the occasion to update this.

maximebiloe replied:
Too late for 1.7.2 :(

LittleBigDev replied:
~~I think so. Should I change this couple of translations in this PR ?~~
Edit : according to Maxime's comment, I won't fix it here.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -3762,44 +3797,66 @@ public function storeContactImportOne($info, $shop_is_feature_active, $regenerat
+        if (!$res) {
+            $this->errors[] = Db::getInstance()->getMsgError() . ' '
+                . sprintf(
                     $this->trans('%1$s (ID: %2$s) cannot be saved', array(), 'Admin.Advparameters.Notification'),
```
While we refacto stuff, it would be great to not use sprintf for a translated string, but insert  `$info['name']` and `($store->id ? $store->id : 'null')` in the second parameter of trans instead

LittleBigDev replied:
In progress...

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -3762,44 +3797,66 @@ public function storeContactImportOne($info, $shop_is_feature_active, $regenerat
+
+        // If store doesn't exist or update failed
+        if (!$res) {
+            $store->force_id = (bool)$forceIds;
```
Seems to be a duplicate of line 3724

LittleBigDev replied:
Yes I forgot to remove it here.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -3762,44 +3797,66 @@ public function storeContactImportOne($info, $shop_is_feature_active, $regenerat
+
+        $res = false;
+        if ($store->storeExists($store->id)) {
+            $res = $store->update();
```
Would `save()` help you? It handles both `update()` and `add()` whether it exists or not.

LittleBigDev replied:
Yes it would. Should I fix this ?

LittleBigDev replied:
I had a look at `ObjectModel::save()` method and behavior is slightly different.
`ObjectModel::save()` only checks if an `id` is present and if so, `ObjectModel::update()` is called.
Our code explicitely checks in DB if this is a known id. I'm ok to use `ObjectModel::save()`, but I need your feedback before.

Quetzacoalt91 reviewed (commented):
Good idea to refacto / clean, all comments are made regarding this task.

However, some changes were not easy to understand. In my opinion, we need more tests prior merge.

aleeks commented:
@LittleBigDev I put your PR in `develop` branch, we can just include this one to `1.7.2.x` now.
Can you create a PR in `1.7.2.x` with just the fix please ?

Regards

LittleBigDev committed:
CO: Refactored placeholders usage in translated keys (removed sprintf)

LittleBigDev committed:
CO: code refactoring + fixed mistakes and refacto oversights

LittleBigDev committed:
CO: code refactoring
- "if" statements arrangement
- SQL query strings arrangement
- php code sniffer warnings

eternoendless commented on classes/Manufacturer.php:
```diff
@@ -64,6 +64,8 @@ class ManufacturerCore extends ObjectModel
     /** @var bool active */
     public $active;
 
+    public $shop;
```
Please don't forget to add PHPDoc if you know what this property contains.

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -118,47 +118,106 @@ public function __construct()
+                            'Attribute (Name:Type:Position)',
+                            array(),
+                            'Admin.Advparameters.Feature'
+                        ) . '*',
```
Symfony standards say:

> Add a single space around binary operators (`==`, `&&`, ...), **with the exception of the concatenation (`.`) operator**;

So unfortunately you need to remove the spaces around the `.` operator.

See: http://symfony.com/doc/current/contributing/code/standards.html#structure

LittleBigDev replied:
Seems like there is no rule in the code (not in core nor modules). Core team opinion seems unclear too. And despite SF standard (no space usage), SF community seems divided too... SF PGP CS Fixer seems kinda buggy as soon as you concatenate on multiple lines (indentation adds spaces before or after the dot (`.`).
And PSR are just ignoring this case.
Well, I guess we have no rules settled yet...

maximebiloe replied:
@eternoendless gave you the link and its pretty clear, so what is unclear?

LittleBigDev replied:
- Around 200 occurrences of "` . `"in /src/ (new code)
- Around 7500 occurrences in the whole project (including generated php code in cache)
- Regular use of spaces by @julienbourdeau, @rGaillard, @thierrymarianne, @antoin-m, @fatmaBouchekoua (should I name them all ?) and community contributors (meaning that you guys validated it).
- Internet debated about if for long. Even at Symfony they tend not to strictly stick to this rule. Other big framework and communities are using spaces around concatenation operator.
- When I tried to talk about it with core devs, I had no clear answer. It seems like everybody tries his best to not break what he sees in the file, file by file, case by case. It also seems like any single people I asked prefers using spaces (including me).

So yes, all this is unclear.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -559,12 +855,14 @@ public function __construct()
 
-        $this->separator = ($separator = Tools::substr(strval(trim(Tools::getValue('separator'))), 0, 1)) ? $separator :  ';';
+        $separator = Tools::substr(strval(trim(Tools::getValue('separator'))), 0, 1);
+        $this->separator = $separator ? $separator :  ';';
```
Watch it, you wrote two spaces after `:`

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -559,12 +855,14 @@ public function __construct()
         $this->convert = false;
-        $this->multiple_value_separator = ($separator = Tools::substr(strval(trim(Tools::getValue('multiple_value_separator'))), 0, 1)) ? $separator :  ',';
+        $separator = Tools::substr(strval(trim(Tools::getValue('multiple_value_separator'))), 0, 1);
+        $this->multiple_value_separator = $separator ? $separator :  ',';
```
Same thing here

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -652,14 +959,18 @@ public function renderForm()
 
         switch ($last) {
+
+            /** @noinspection PhpMissingBreakStatementInspection */
```
You shouldn't need this if your case ends with a comment

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -679,7 +990,7 @@ public function renderForm()
             'id_language' => ($id_lang_selected) ? $id_lang_selected : $this->context->language->id,
             'available_fields' => $this->getAvailableFields(),
-            'truncateAuthorized' => (Shop::isFeatureActive() && $this->context->employee->isSuperAdmin()) || !Shop::isFeatureActive(),
+            'truncateAuthorized' => !(Shop::isFeatureActive()) || $this->context->employee->isSuperAdmin(),
```
You can drop the parentheses:

```php
!Shop::isFeatureActive()
```

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -822,7 +1179,9 @@ class="type_value">
             foreach ($line as $nb_c => $column) {
-                if ((MAX_COLUMNS * (int)$current_table <= $nb_c) && ((int)$nb_c < MAX_COLUMNS * ((int)$current_table + 1))) {
+                if ((MAX_COLUMNS * (int)$current_table <= $nb_c)
+                    && ((int)$nb_c < MAX_COLUMNS * ((int)$current_table + 1))
```
You should add a single space between the type casting operator and the expression to cast:

```php
(int) $nb_c
```

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1031,16 +1402,25 @@ protected static function setEntityDefaultValues(&$entity)
-        if (isset(self::$validators[$key][1]) && self::$validators[$key][1] == 'createMultiLangField' && Tools::getValue('iso_lang')) {
-            $id_lang = Language::getIdByIso(Tools::getValue('iso_lang'));
-            $tmp = call_user_func(self::$validators[$key], $infos);
+        $validator = self::$validators[$key];
```
Are you sure that `isset(self::$valodators[$key])`?

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1031,16 +1402,25 @@ protected static function setEntityDefaultValues(&$entity)
             }
-        } elseif (!empty($infos) || $infos == '0') { // ($infos == '0') => if you want to disable a product by using "0" in active because empty('0') return true
-                $entity->{$key} = isset(self::$validators[$key]) ? call_user_func(self::$validators[$key], $infos) : $infos;
+        } elseif (!empty($infos) || $infos == '0') { // Because "0" string is not an empty value
```
Please use triple equal, or `(int) 0` and `false` will pass your check

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1031,16 +1402,25 @@ protected static function setEntityDefaultValues(&$entity)
-        } elseif (!empty($infos) || $infos == '0') { // ($infos == '0') => if you want to disable a product by using "0" in active because empty('0') return true
-                $entity->{$key} = isset(self::$validators[$key]) ? call_user_func(self::$validators[$key], $infos) : $infos;
+        } elseif (!empty($infos) || $infos == '0') { // Because "0" string is not an empty value
+            $entity->{$key} = isset($validator)
```
So `$validator` may not have been set! (see my comment at the top of this method)

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1190,21 +1594,53 @@ protected static function copyImg($id_entity, $id_image = null, $url = '', $enti
+     *
+     * @param $tgtWidth
+     * @param $tgtHeight
+     * @param $pathInfos
```
Please add types and description if you know them

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1190,21 +1594,53 @@ protected static function copyImg($id_entity, $id_image = null, $url = '', $enti
+     *
+     * @return string
+     */
+    protected static function getBestPath($tgtWidth, $tgtHeight, $pathInfos)
```
Since you renamed the variable, please try not to use "infos". The word "information" has no plural form in english, and it makes my eyes hurt 😛

LittleBigDev replied:
Fixed in next commit (and everywhere in this class).

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1190,21 +1594,53 @@ protected static function copyImg($id_entity, $id_image = null, $url = '', $enti
 
-    public function categoryImport($offset = false, $limit = false, &$crossStepsVariables = false, $validateOnly = false)
-    {
+    public function categoryImport(
```
Don't forget to add a PHPDoc header whenever possible

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1224,19 +1660,25 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
-        $line_count = 0;
-        for ($current_line = 0; ($line = fgetcsv($handle, MAX_LINE_SIZE, $this->separator)) && (!$limit || $current_line < $limit); $current_line++) {
-            $line_count++;
+        $lineCount = 0;
```
Since you changed this one, please change all the other local variables to camelCase.

LittleBigDev replied:
Fixed.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1260,26 +1702,42 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
 
-        if ($force_ids && isset($info['id']) && (int)$info['id']) {
-            $category = new Category((int)$info['id']);
+        if ($forceIds && isset($categoryId) && $categoryId) {
```
I think you can change
```php
isset($categoryId) && $categoryId
```
by
```php
0 < $categoryId
```

LittleBigDev replied:
Fixed in next commit.

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1260,26 +1702,42 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
         } else {
-            if (isset($info['id']) && (int)$info['id'] && Category::existsInDatabase((int)$info['id'], 'category')) {
-                $category = new Category((int)$info['id']);
+            if (isset($categoryId) && $categoryId && Category::existsInDatabase($categoryId, 'category')) {
```
This check is already done in line [#1736](https://github.com/PrestaShop/PrestaShop/pull/8122/files#diff-7bdf6b540c21355166c0d726ffb9bf81R1736):
```php
isset($categoryId) && $categoryId
```
I suggest making that check earlier.

LittleBigDev replied:
Fixed in next commit.

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1290,16 +1748,19 @@ protected function categoryImportOne($info, $default_language_id, $id_lang, $for
             // Validation for parenting itself
-            if ($validateOnly && ($category->parent == $category->id) || (isset($info['id']) && $category->parent == (int)$info['id'])) {
+            if ($validateOnly && $category->parent == $category->id
+                || isset($categoryId) && $category->parent == $categoryId
```
Again, since $categoryId is guaranteed to be an integer, it will always be set.

LittleBigDev replied:
Fixed in next commit.

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1312,109 +1773,125 @@ protected function categoryImportOne($info, $default_language_id, $id_lang, $for
-                            $category_to_create->name[$id_lang],
-                            (isset($category_to_create->id) && !empty($category_to_create->id))? $category_to_create->id : 'null'
-                        );
+                        $this->trans('%1$s (ID: %2$s) cannot be saved', array(
```
I think you accidentally removed `$this->errors[] =`

LittleBigDev replied:
indeed

LittleBigDev replied:
indeed

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1312,109 +1773,125 @@ protected function categoryImportOne($info, $default_language_id, $id_lang, $for
                 '%1$s' => $bak,
-                '%2$s' => (isset($info['id']) && !empty($info['id']))? $info['id'] : 'null',
-                '%3$s' => $category->link_rewrite[$default_language_id],
+                '%2$s' => (isset($categoryId) && !empty($categoryId))? $categoryId : 'null',
```
Again, just check against zero

LittleBigDev replied:
fixed in next commit

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1312,109 +1773,125 @@ protected function categoryImportOne($info, $default_language_id, $id_lang, $for
-                !$validateOnly) {
+            if ($category->id
+                && $category->categoryExists($category->id)
+                && !in_array($category->id, $baseCategories)
```
Warning! The order of `$categories_home_root` was inverted compared to `$baseCategories`.

$tab_categ: (--> [see code](https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/admin/AdminImportController.php?utf8=%E2%9C%93#L1271))
```php
$tab_categ = array(Configuration::get('PS_HOME_CATEGORY'), Configuration::get('PS_ROOT_CATEGORY'));
```

$categories_home_root: (--> [see code](https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/admin/AdminImportController.php?utf8=%E2%9C%93#L1406))
```php
$categories_home_root = array(Configuration::get('PS_ROOT_CATEGORY'), Configuration::get('PS_HOME_CATEGORY'));
```

The indexes won't match. So either that was a bug and this fixes it, or you're introducing a regression.

LittleBigDev replied:
Indexes are not used here. Just a plain in_array() to check current (soon to be created) category id against known existing categories ids.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1473,11 +1953,21 @@ protected function categoryImportOne($info, $default_language_id, $id_lang, $for
+        $moreStep = 0
+    ) {
+        if (!is_array($crossStepsVariables)) {
+            $crossStepsVariables = array();
```
This variable is passed by reference. I hope not, but there may be code depending on its value staying `false`

LittleBigDev replied:
Only variables can be passed by reference. It's impossible to fall back on the default value unless a `null` value variable is passed. Core and native modules don't do that. And there is no way such a case can happen somewhere else. We figured this out with Aurélien (@toutantic).
`$crossStepsVariables` is ment to be an array, and any other value would just be a mishap.

Quetzacoalt91 replied:
I'm okay with the suggested change.

But I wonder if `array &$crossStepsVariables = array(),` would work as well and would fit better to our need.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1539,15 +2035,18 @@ public function productImport($offset = false, $limit = false, &$crossStepsVaria
     }
 
-    protected function productImportAccessories($offset, $limit, &$crossStepsVariables)
+    protected function productImportAccessories($offset, $limit, &$crossStepsVariables = array())
```
Changing the method's signature may break overrides

LittleBigDev replied:
That's true. I'll revert this.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1539,15 +2035,18 @@ public function productImport($offset = false, $limit = false, &$crossStepsVaria
     {
-        if ($crossStepsVariables === false || !array_key_exists('accessories', $crossStepsVariables)) {
+        if (!is_array($crossStepsVariables)) {
+            $crossStepsVariables = array();
```
`$crossStepsVariables` is passed by reference, so writing on it here changes the outcome.

LittleBigDev replied:
Same comment [as above](https://github.com/PrestaShop/PrestaShop/pull/8122#discussion_r128494210)

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1575,8 +2074,10 @@ protected function productImportAccessories($offset, $limit, &$crossStepsVariabl
-            $accessories[$product_id] = 0; // In JSON, 0 is lighter than null or false
+            // Empty value to reduce array weight (that goes through HTTP requests each time).
+            // But do not unset array entry!
+            // In JSON, 0 is lighter than null or false
```
Lighter? I think the author probably meant "lower".

LittleBigDev replied:
Lighter in term of memory usage / character length / weight within http request.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1594,24 +2095,36 @@ protected function productImportAccessories($offset, $limit, &$crossStepsVariabl
+        if ($forceIds && $productId) {
+            $product = new Product($productId);
+        } elseif ($matchRef && array_key_exists('reference', $info)) {
+            $datas = Db::getInstance()->getRow(
```
Another word to avoid: "datas". As "information", data has no plural form in english.
It would be better to use a more semantic wording, like "productData"

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1594,24 +2095,36 @@ protected function productImportAccessories($offset, $limit, &$crossStepsVariabl
             }
-        } elseif (array_key_exists('id', $info) && (int)$info['id'] && Product::existsInDatabase((int)$info['id'], 'product')) {
-            $product = new Product((int)$info['id']);
+        } elseif (array_key_exists('id', $info) && $productId && Product::existsInDatabase($productId, 'product')) {
```
You can remove `array_key_exists('id', $info)`, as that is verified by `$productId > 0`

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1634,13 +2147,13 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
         AdminImportController::arrayWalk($info, array('AdminImportController', 'fillInfo'), $product);
 
-        if (!$shop_is_feature_active) {
+        if (!$shopIsFeatureActive) {
```
I think you can merge this if/else block with the one just after it.

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1666,42 +2179,57 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
+        ) {
             $product->id_manufacturer = (int)$product->manufacturer;
-        } elseif (isset($product->manufacturer) && is_string($product->manufacturer) && !empty($product->manufacturer)) {
+        } elseif (isset($product->manufacturer)
```
Since this check is done twice, you could move it up.

LittleBigDev replied:
Fixed in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1666,42 +2179,57 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
-                    !$validateOnly && // Do not move this condition: previous tests should be played always, but next ->add() test should not be played in validateOnly mode
-                    $manufacturer->add()) {
+                // FIXME needing such a comment about "&& !$validateOnly" position is a code smell. Refacto needed.
+                if (($field_error = $manufacturer->validateFields(UNFRIENDLY_ERROR, true)) === true
```
Yoda statements, you should use.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1760,30 +2290,46 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
+                        $category_to_create->id_parent = Configuration::get('PS_HOME_CATEGORY');
+                        $category_link_rewrite = Tools::link_rewrite($category_to_create->name[$defaultLanguageId]);
+
+
```
No more than one empty line please 🙂

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1875,28 +2428,29 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
+            if ($matchRef && $product->reference && $product->existsRefInDatabase($product->reference)) {
+                $datas = Db::getInstance()->getRow(
+                    'SELECT product_shop.`date_add`, p.`id_product` '
+                    . 'FROM `' . _DB_PREFIX_ . 'product` p ' . Shop::addSqlAssociation('product', 'p')
```
Please add a leading space before FROM, so there will be less of a risk of someone forgetting to leave a trailing space on the previous line 🙂

LittleBigDev replied:
Fixed everywhere in next commit.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1875,28 +2428,29 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
-					WHERE p.`id_product` = '.(int)$product->id, false);
+                $datas = Db::getInstance()->getRow(
+                    'SELECT product_shop.`date_add` '
+                    . 'FROM `' . _DB_PREFIX_ . 'product` p ' . Shop::addSqlAssociation('product', 'p')
```
Same here, leading space in front of FROM

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1943,28 +2497,40 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
+            $this->errors[] = $this->trans('%1$s (ID: %2$s) cannot be saved', array(
+                !empty($info['name']) ? Tools::safeOutput($info['name']) : 'No Name',
+                !empty($productId) ? Tools::safeOutput($productId) : 'No ID',
+            ), 'Admin.Advparameters.Notification');
```
Please break `$this->trans(...)` into one argument per line.

LittleBigDev replied:
ok

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1943,28 +2497,40 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
-                $id_product_supplier = (int)ProductSupplier::getIdByProductAndSupplier((int)$product->id, 0, (int)$product->id_supplier);
+            if (!$validateOnly
+                && isset($product->id)
+                && $product->id
```
I guess you can test `!empty($product->id)` instead?

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -1943,28 +2497,40 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
+            if (!$validateOnly
+                && isset($product->id)
+                && $product->id
+                && isset($product->id_supplier) && property_exists($product, 'supplier_reference')
```
Either all in one line, or one per line, but not both please 😉

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -2221,32 +2857,54 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
+            if (isset($product->accessories)
+                && !$validateOnly
+                && is_array($product->accessories)
+                && count($product->accessories)
```
`0 < count($product->accessories)` for clarity

LittleBigDev replied:
ok

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -2263,39 +2921,49 @@ public function productImportCreateCat($default_language_id, $category_name, $id
         $category_to_create->name = AdminImportController::createMultiLangField(trim($category_name));
         $category_to_create->active = 1;
-        $category_to_create->id_parent = (int)$id_parent_category ? (int)$id_parent_category : (int)Configuration::get('PS_HOME_CATEGORY'); // Default parent is home for unknown category to create
+        $category_to_create->id_parent = (int)$id_parent_category
```
`(0 < (int) $id_parent_category)` for clarity

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -2263,39 +2921,49 @@ public function productImportCreateCat($default_language_id, $category_name, $id
+        $validateOnly = false
+    ) {
+        if (!is_array($crossStepsVariables)) {
+            $crossStepsVariables = array();
```
Same thing, variable is passed by reference, writing on it now may break expectations.

LittleBigDev replied:
Same comment [as above](https://github.com/PrestaShop/PrestaShop/pull/8122#discussion_r128494210)

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -2375,12 +3056,12 @@ protected function attributeImportOne($info, $default_language, &$groups, &$attr
-			', false);
+            $datas = Db::getInstance()->getRow(
+                'SELECT p.`id_product` '
+                . 'FROM `' . _DB_PREFIX_ . 'product` p ' . Shop::addSqlAssociation('product', 'p')
```
Leading space before FROM

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -3682,30 +4606,49 @@ public function storeContactImport($offset = false, $limit = false, $validateOnl
+     * @param       $notUsed      : Not used anymore.
+     * @param bool  $regenerate   : If images should be regenerated
+     * @param bool  $forceIds     : If passed store id should be used for store creation
+     * @param bool  $validateOnly : If set to true, store will not be updated nor created with $info data
```
When writing PHPDoc, please **don't** use a separator between the variable name and the description

LittleBigDev replied:
Fixed

LittleBigDev committed:
CO: PSR compliancy (variables naming)
I didn't refacto the visible symbols (public, static...)

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -3725,23 +4668,38 @@ public function storeContactImportOne($info, $shop_is_feature_active, $regenerat
-                        $this->errors[] = ($field_error !== true ? $field_error : '').(isset($lang_field_error) && $lang_field_error !== true ? $lang_field_error : '').
-                            Db::getInstance()->getMsgError();
+                    if ($fieldsValidationResult !== true
+                        || isset($langFieldsValidationResult) && $langFieldsValidationResult !== true
```
Please wrap `isset($langFieldsValidationResult) && $langFieldsValidationResult !== true` with parentheses for clarity.

LittleBigDev replied:
- && is prevalent on ||
- I created a new line for this group
- but ok, I'll change this in next commit.

LittleBigDev replied:
Done.

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -4078,7 +5147,8 @@ protected function supplyOrdersDetailsImportOne($info, &$products, &$reset, $for
 				');
                 $query->from('product', 'p');
-                $query->leftJoin('product_attribute', 'pa', 'pa.id_product = p.id_product AND id_product_attribute = '.(int)$id_product_attribute);
+                $query->leftJoin('product_attribute', 'pa', 'pa.id_product = p.id_product AND id_product_attribute = '
```
If the line is too long, I'd prefer if it was broken up by arguments (one per line).

eternoendless commented on controllers/admin/AdminImportController.php:
```diff
@@ -4224,7 +5294,8 @@ protected function truncateTables($case)
                 Db::getInstance()->execute('TRUNCATE TABLE `'._DB_PREFIX_.'product_carrier`');
                 Db::getInstance()->execute('TRUNCATE TABLE `'._DB_PREFIX_.'cart_product`');
-                if (count(Db::getInstance()->executeS('SHOW TABLES LIKE \''._DB_PREFIX_.'favorite_product\' '))) { //check if table exist
+                //check if table exist before truncate
```
"table _exists_" 😉

eternoendless reviewed (dismissed):
Thanks for this refacto! There are many little things to fix :)

aleeks commented on classes/Product.php:
```diff
@@ -52,6 +52,9 @@ class ProductCore extends ObjectModel
     /** @var int default Category id */
     public $id_category_default;
 
+    /** @var array */
```
Hello @LittleBigDev 
I've no idea of the context, just an alert, is it really array ?

LittleBigDev replied:
Yes it is. Have a look at [this line in AdminImportController](https://github.com/PrestaShop/PrestaShop/blob/ecaf9938fb0a455f72a6d3ba82849bb738c17761/controllers/admin/AdminImportController.php#L1753).

LittleBigDev committed:
Fixed PR according to @eternoendless advice

LittleBigDev commented on controllers/admin/AdminImportController.php:
```diff
@@ -1290,16 +1748,19 @@ protected function categoryImportOne($info, $default_language_id, $id_lang, $for
             }
-            if (isset($cat_moved[$category->parent])) {
-                $category->parent = $cat_moved[$category->parent];
+            if (isset($catMoved[$category->parent])) {
```
fixed in next commit

LittleBigDev commented on classes/Product.php:
```diff
@@ -52,6 +52,9 @@ class ProductCore extends ObjectModel
     /** @var int default Category id */
     public $id_category_default;
 
+    /** @var array */
```
Yes it is. Look at usages in ImportAdminController

maximebiloe commented on controllers/admin/AdminImportController.php:
```diff
@@ -339,23 +547,52 @@ public function __construct()
-                        'help' => $this->trans('Ignore this field if you don\'t use the Multistore tool. If you leave this field empty, the default shop will be used.', array(), 'Admin.Advparameters.Help'),
+                        'help' => $this->trans(
+                            'Ignore this field if you don\'t use the Multistore tool. If you leave this field empty, '
+                                . 'the default shop will be used.',
```
You can't split a translatable string like that, the parser will fail.
Please, remove the concatenation even if you did it for the line length.

LittleBigDev replied:
This would break the standards. @eternoendless please validate this.

aleeks replied:
It's not about standards here ..It's about code that works :-1:

maximebiloe replied:
If you think standards are more important than not breaking things...

LittleBigDev replied:
Then maybe we are debating about the wrong subject then.
Ask me to break the standards and I'll do it.

maximebiloe commented on controllers/admin/AdminImportController.php:
```diff
@@ -420,19 +667,28 @@ public function __construct()
                         'label' => $this->trans('ID / Name of group shop', array(), 'Admin.Advparameters.Feature'),
-                        'help' => $this->trans('Ignore this field if you don\'t use the Multistore tool. If you leave this field empty, the default shop will be used.', array(), 'Admin.Advparameters.Help'),
+                        'help' => $this->trans(
+                            'Ignore this field if you don\'t use the Multistore tool. If you leave this field empty, '
```
You can't split a translatable string like that, the parser will fail.
Please, remove the concatenation even if you did it for the line length.

maximebiloe commented on controllers/admin/AdminImportController.php:
```diff
@@ -480,17 +740,28 @@ public function __construct()
                         'label' => $this->trans('ID / Name of shop', array(), 'Admin.Advparameters.Feature'),
-                        'help' => $this->trans('Ignore this field if you don\'t use the Multistore tool. If you leave this field empty, the default shop will be used.', array(), 'Admin.Advparameters.Help'),
+                        'help' => $this->trans(
+                            'Ignore this field if you don\'t use the Multistore tool. If you leave this field empty, '
```
You can't split a translatable string like that, the parser will fail.
Please, remove the concatenation even if you did it for the line length.

maximebiloe commented on controllers/admin/AdminImportController.php:
```diff
@@ -559,187 +862,270 @@ public function __construct()
-					<code>php_value post_max_size 20M</code> '.
-                    $this->trans('(click to open "Generators" page)', array(), 'Admin.Advparameters.Notification').'</a>';
+                    $uploadErrorMessage = $this->trans(
+                        'The uploaded file exceeds the post_max_size directive in php.ini. If your server '
```
You can't split a translatable string like that, the parser will fail.
Please, remove the concatenation even if you did it for the line length.

LittleBigDev committed:
CO: Reverted some refactoring for the sake of translations parser ("trans(" and the whole translated string need to be on the same line).
PSR broken rules are intended.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -559,187 +824,259 @@ public function __construct()
+            'id_language' => ($idLangSelected) ? $idLangSelected : $this->context->language->id,
             'available_fields' => $this->getAvailableFields(),
-            'truncateAuthorized' => (Shop::isFeatureActive() && $this->context->employee->isSuperAdmin()) || !Shop::isFeatureActive(),
+            'truncateAuthorized' => !(Shop::isFeatureActive()) || $this->context->employee->isSuperAdmin(),
```
Parentheses around `Shop::isFeatureActive()` can be removed

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1260,161 +1805,180 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
+                || $category->parent == $categoryId
+            ) {
+                $this->errors[] = $this->trans('The category ID must be unique. It can\'t be the same as the one for the parent category (ID: %1$s).',
+                    array($categoryId ? $categoryId : 'null'),
```
When using a placeholder with the Symfony translator, your must name them (https://symfony.com/doc/current/components/translation/usage.html#message-placeholders). Then you affect this name a variable to be replaced in the translated string.

Here is the correct syntax to translate this string:
```php
$this->errors[] = $this->trans('The category ID must be unique. It can\'t be the same as the one for the parent category (ID: %id%).',
    array('%id%' => ($categoryId ? $categoryId : 'null')),
    'Admin.Advparameters.Notification'
);
```

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1260,161 +1805,180 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
             if ($validateOnly && isset($category->name) && ($category->parent == $category->name)) {
-                $this->errors[] = $this->trans(
-                    'A category can\'t be its own parent. You should rename it (current name: %1$s).',
+                $this->errors[] = $this->trans('A category can\'t be its own parent. You should rename it (current name: %1$s).',
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1260,161 +1805,180 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
-                        );
+                        $this->errors[] = $this->trans('%1$s (ID: %2$s) cannot be saved', array(
+                            $categoryToCreate->name[$idLang],
+                            !empty($categoryToCreate->id) ? $categoryToCreate->id : 'null',
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1260,161 +1805,180 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
                 $category->link_rewrite = 'friendly-url-autogeneration-failed';
-                $this->warnings[] = sprintf($this->trans('URL rewriting failed to auto-generate a friendly URL for: %s', array(), 'Admin.Advparameters.Notification'), $category->name[$default_language_id]);
+                $this->warnings[] = $this->trans('URL rewriting failed to auto-generate a friendly URL for: %s',
+                    array($category->name[$defaultLanguageId]),
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1260,161 +1805,180 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
-                '%2$s' => (isset($info['id']) && !empty($info['id']))? $info['id'] : 'null',
-                '%3$s' => $category->link_rewrite[$default_language_id],
+                '%2$s' => $categoryId ? $categoryId : 'null',
+                '%3$s' => $category->link_rewrite[$defaultLanguageId],
```
This is the perfect moment to update the variable names to make them more explicit. If needed you can check with @AlexEven

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1260,161 +1805,180 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
-                    $this->trans('A category cannot be its own parent. The parent category ID is either missing or unknown (ID: %1$s).', array(), 'Admin.Advparameters.Notification'),
-                    (isset($info['id']) && !empty($info['id']))? $info['id'] : 'null'
+                $this->errors[] = $this->trans('A category cannot be its own parent. The parent category ID is either missing or unknown (ID: %1$s).',
+                    array(!empty($categoryId) ? $categoryId : 'null'),
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1260,161 +1805,180 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
+                // Default parent is home for unknown category to create :
+                $categoryToCreate->id_parent = Configuration::get('PS_HOME_CATEGORY');
+
+                // FIXME needing such a comment about "&& !$validateOnly" position is a code smell. Refacto needed.
```
![i-heard-you-like-refacto-so-i-refacto-so-you-could-know-where-to-refacto](https://user-images.githubusercontent.com/6768917/29885221-fb268480-8dad-11e7-90a8-67f68abb2abf.jpg)

:laughing:

LittleBigDev replied:
Yeah, it was going beyond the "tiny" refacto I was doing. But I wanted to leave a comment, not to forget later....

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1428,32 +1992,34 @@ protected function categoryImportOne($info, $default_language_id, $id_lang, $for
         if (isset($category->image) && !empty($category->image)) {
             if (!(AdminImportController::copyImg($category->id, null, $category->image, 'categories', !$regenerate))) {
-                $this->warnings[] = $category->image.' '.$this->trans('cannot be copied.', array(), 'Admin.Advparameters.Notification');
+                $this->warnings[] = $category->image.' '.$this->trans('cannot be copied.',
```
In other languages, the image name may go at a different place. We should add it as a parameter.

AlexEven replied:
+1, thanks @Quetzacoalt91 for spotting this!

LittleBigDev replied:
Maybe this is outside the scope of this refacto ?

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1428,32 +1992,34 @@ protected function categoryImportOne($info, $default_language_id, $id_lang, $for
+            $this->errors[] = $this->trans('%1$s (ID: %2$s) cannot be %3$s', array(
+                !empty($info['name']) ? Tools::safeOutput($info['name']) : 'No Name',
+                !empty($categoryId) ? Tools::safeOutput($categoryId) : 'No ID',
+                $validateOnly ? 'validated' : 'saved',
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1428,32 +1992,34 @@ protected function categoryImportOne($info, $default_language_id, $id_lang, $for
+                );
 
-                if (!$shop_is_feature_active) {
+                if (!$shopIsFeatureActive) {
```
Never true, can be removed

LittleBigDev replied:
So, it looks like I should remove the whole block from line 2015 to line 2038 (and same thing wherever I find a test depending on `$shopIsFeatureActive` value).
Am I right ?

Quetzacoalt91 replied:
No, that's too much. ;)
But in that case, we have:

```php
if ($shopIsFeatureActive) {
    [...]
    if (!$shopIsFeatureActive) {
    }
}
```
Which cannot happen.

Removing this test and the line after (`$info['shop'] = 1;`) is enough

LittleBigDev replied:
Will do.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1569,80 +2153,88 @@ protected function productImportAccessories($offset, $limit, &$crossStepsVariabl
+        if ($matchRef && !empty($productData['reference'])) {
+            $data = Db::getInstance()->getRow(
+                'SELECT p.`id_product`'
+                . ' FROM `' . _DB_PREFIX_ . 'product` p ' . Shop::addSqlAssociation('product', 'p')
```
Seen while reviewing #8255, I guess `Shop::addSqlAssociation('product', 'p')` is useless

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1659,49 +2251,61 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
+                        $manufacturer->associateTo($product->id_shop_list);
+                    } else {
+                        if (!$validateOnly) {
+                            $this->errors[] = $this->trans('%1$s (ID: %2$s) cannot be saved', [
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 replied:
Use long array syntax to create your arrays please.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1711,23 +2315,25 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
                     if (!$validateOnly) {
-                        $this->errors[] = sprintf(
-                            $this->trans('%1$s (ID: %2$s) cannot be saved', array(), 'Admin.Advparameters.Notification'),
+                        $this->errors[] = $this->trans('%1$s (ID: %2$s) cannot be saved', array(
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1756,39 +2362,54 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
-                                    $category_to_create->name[$default_language_id],
-                                    (isset($category_to_create->id) && !empty($category_to_create->id))? $category_to_create->id : 'null'
-                                );
+                                $this->errors[] = $this->trans('%1$s (ID: %2$s) cannot be saved', array(
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1756,39 +2362,54 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
                     } else {
-                        $this->errors[] = $this->trans(
-                            '%s cannot be saved',
+                        $this->errors[] = $this->trans('%s cannot be saved',
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -2049,24 +2708,27 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
-                                'Invalid tag(s) (%s)',
+                        $isTagAdded = Tag::addTags($key, $product->id, $str, $this->multiple_value_separator);
+                        if (!$isTagAdded) {
+                            $this->addProductWarning(Tools::safeOutput($productData['name']), (int)$product->id, $this->trans('Invalid tag(s) (%s)',
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -2094,19 +2756,30 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
+                            )) {
                                 $image->delete();
-                                $this->warnings[] = sprintf($this->trans('Error copying image: %s', array(), 'Admin.Advparameters.Notification'), $url);
+                                $this->warnings[] = $this->trans('Error copying image: %s',
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -2116,7 +2789,10 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
 
                     if ($error) {
-                        $this->warnings[] = sprintf($this->trans('Product #%1$d: the picture (%2$s) cannot be saved.', array(), 'Admin.Advparameters.Notification'), $image->id_product, $url);
+                        $this->warnings[] = $this->trans('Product #%1$d: the picture (%2$s) cannot be saved.', array(
```
Same comment about translation placeholder. Please use a variable name surrounded by `%%` an add it in the second parameter.

Quetzacoalt91 reviewed (changes requested):
It stopped during the review to avoid spam.

It would be great to give names to the placeholders in translated strings. This is important to give a context to the translator, and may help us to remove the duplicated strings (started in #7840).

AlexEven commented on controllers/admin/AdminImportController.php:
```diff
@@ -379,28 +582,38 @@ public function __construct()
-                    'lastname' => array('label' => $this->trans('Last name', array(), 'Admin.Global').'*'),
-                    'firstname' => array('label' => $this->trans('First name ', array(), 'Admin.Global').'*'),
-                    'address1' => array('label' => $this->trans('Address', array(), 'Admin.Global').'*'),
+                    'lastname' => array('label' => $this->trans('Last name', array(), 'Admin.Global') . '*'),
```
How does it behave in RTL?

LittleBigDev replied:
Maybe @maximebiloe could answer this ?

Quetzacoalt91 replied:
Not as expected. I found several part of text which needs to be updated (in another PR).

AlexEven commented on controllers/admin/AdminImportController.php:
```diff
@@ -3405,29 +4318,41 @@ protected function manufacturerImportOne($info, $shop_is_feature_active, $regene
+                    !$regenerate
+                )) {
+                    // TODO : insert translation placeholder here
+                    $this->warnings[] = $manufacturer->image . ' ' . $this->trans('cannot be copied.',
```
We need a variable here too

AlexEven commented on controllers/admin/AdminImportController.php:
```diff
@@ -3677,37 +4642,54 @@ public function storeContactImport($offset = false, $limit = false, $validateOnl
+            $imgWasCopied = AdminImportController::copyImg($store->id, null, $store->image, 'stores', !$regenerate);
+            if (!$imgWasCopied) {
+                $this->warnings[] = $store->image . ' '
+                    . $this->trans('cannot be copied.', array(), 'Admin.Advparameters.Notification');
```
This needs to be a full sentence with a variable.

LittleBigDev committed:
BO: AdminImportController refacto fixes after review.

LittleBigDev commented:
Pushed fixes according to review comments.

LittleBigDev committed:
BO: AdminImportController refacto fixes after review.

antoin-m commented:
![image](https://thumbs.gfycat.com/InsistentSpecificGemsbok-size_restricted.gif)

xBorderie commented:
![1vbiaa](https://user-images.githubusercontent.com/2833749/30064635-a4f3aed6-9252-11e7-9e40-0b6aa3ce73d0.jpg)

mickaelandrieu commented:
Hello @LittleBigDev,

this need to be rebased (and probably validated again by  @Quetzacoalt91 ?)

Mickaël

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -559,187 +824,259 @@ public function __construct()
+                        . $generatorsPageMessage
+                        . '</a>';
+
+                    $_FILES['file']['error'] = $uploadErrorMessage . $directiveExample;
```
Ach, I didn't see this line. For the RTL languages, this will need to be merged in a single string to translate (in another PR).

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -774,6 +1111,11 @@ public function renderView()
+    /**
+     * Initializes import toolbar (for import display only)
+     *
+     * @return void
```
Line to be removed

LittleBigDev replied:
Will do

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1028,228 +1473,328 @@ protected static function setEntityDefaultValues(&$entity)
+        // If nothing to fill, just get out ("0" string is NOT nothing).
+        if (empty($data) && $data !== '0') {
+            return true;
+        }
```
It seems the first check was made before to handle the multilang case, and now this is an check which returns if the value is empty or equals 0.

Did you test that part? I'm afraid that would have some consequence on the feature.

LittleBigDev replied:
I xDebugged it, and in any case, when data is empty, nothing is filled

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1260,161 +1805,180 @@ public function categoryImport($offset = false, $limit = false, &$crossStepsVari
+            Configuration::get('PS_HOME_CATEGORY'),
+            Configuration::get('PS_ROOT_CATEGORY')
+        );
+        $categoryId     = !empty($info['id']) ? (int)$info['id'] : 0;
```
By looking at the next lines of this function, and the calls to new Category, I wonder if you should not set `null` if the $info['id'] is empty.

LittleBigDev replied:
I agree

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -1756,42 +2362,55 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
+                        ) {
+                            $categoryToCreate->link_rewrite = AdminImportController::createMultiLangField(
+                                $categoryLinkRewrite
+                            );
```
Why is it now in the test?

Quetzacoalt91 commented on controllers/admin/AdminImportController.php:
```diff
@@ -3981,120 +5060,147 @@ public function supplyOrdersDetailsImport($offset = false, $limit = false, &$cro
+                $this->errors[] = $this->trans('Tax rate (%rate%) is not valid (at line %line%).', array(
+                    '%rate%' => $taxRate,
+                    '%line%' => $currentLine + 1,
                     $this->trans('Format: Between 0 and 100', array(), 'Admin.Advparameters.Notification')
```
In both case, it seems `$this->trans('Format: Between 0 and 100', array(), 'Admin.Advparameters.Notification')` is never displayed

LittleBigDev committed:
Fixed rebase conflicts

LittleBigDev committed:
Fixed rebase conflicts

LittleBigDev commented:
Hi @Quetzacoalt91 , I tried to address all your comments but after the rebase, I wasn't able to find the related code and/or to understand your suggestions. Could you please check it again ?

EDIT : and maybe we should use Reviewable for such a big review ?

Quetzacoalt91 commented:
Related code can still be found if you do not change anything around the comment.
